src/crimson/net/SocketConnection.cc

```diff
@@ -66,9 +65,32 @@ SocketConnection::get_messenger() const {
   return &messenger;
 }
 
-bool SocketConnection::is_connected()
+seastar::future<bool> SocketConnection::is_connected()
 {
-  return !send_ready.failed();
+  return seastar::smp::submit_to(shard_id(), [this] {
+      return !send_ready.failed();
+    });
+}
+
+seastar::future<> SocketConnection::send(MessageRef msg)
```

Comment:
Argh, we need to take care about RPCing arguments as well. IIURC `MessageRef` is defined in `src/msg/MessageRef.h` as:
```cpp
typedef boost::intrusive_ptr<Message> MessageRef
```

Following the inheritance chain:
```cpp
class Message : public RefCountedObject {
 // ...
};
```
```cpp
struct RefCountedObject {
private:
  mutable std::atomic<uint64_t> nref;
  // ...
};

// ...

static inline void intrusive_ptr_add_ref(const RefCountedObject *p) {
  p->get();
}
static inline void intrusive_ptr_release(const RefCountedObject *p) {
  p->put();
}
```

Still, it looks the deallocation can be performed on another engine than the message was created on :-(.

Reply:
From `src/seastar/include/seastar/core/memory.hh`:
```cpp
/// The seastar memory allocator splits system memory into a pool per
/// logical core (lcore).  Memory allocated one an lcore should be freed
/// on the same lcore; failing to do so carries a severe performance
/// penalty.
```

From the corresponding `.cc`:
```cpp
#ifndef SEASTAR_DEFAULT_ALLOCATOR
// ...

[[gnu::visibility("default")]]
void* operator new(size_t size) {
    trigger_error_injector();
    if (size == 0) {
        size = 1;
    }
    return throw_if_null(allocate(size));
}

[[gnu::visibility("default")]]
void* operator new[](size_t size) {
    trigger_error_injector();
    if (size == 0) {
        size = 1;
    }
    return throw_if_null(allocate(size));
}
#else

// ...

#endif
```

It looks that, at least in the unit tests, we're using the SeaStar's allocator:
```
$ objdump -dC ./bin/unittest_seastar_monc
...
0000000000b36b00 <operator new(unsigned long)>:
  b36b00:       48 83 ec 08             sub    $0x8,%rsp
  b36b04:       b8 01 00 00 00          mov    $0x1,%eax
  b36b09:       48 85 ff                test   %rdi,%rdi
  b36b0c:       48 0f 44 f8             cmove  %rax,%rdi
  b36b10:       e8 2b fe ff ff          callq  b36940 <seastar::memory::allocate(unsigned long)>
  b36b15:       48 85 c0                test   %rax,%rax
  b36b18:       74 05                   je     b36b1f <operator new(unsigned long)+0x1f>
  b36b1a:       48 83 c4 08             add    $0x8,%rsp
  b36b1e:       c3                      retq   
  b36b1f:       e8 1c d1 ff ff          callq  b33c40 <throw_if_null(void*) [clone .part.8]>
  b36b24:       66 90                   xchg   %ax,%ax
  b36b26:       66 2e 0f 1f 84 00 00    nopw   %cs:0x0(%rax,%rax,1)
  b36b2d:       00 00 00 

0000000000b36b30 <operator new[](unsigned long)>:
  b36b30:       eb ce                   jmp    b36b00 <operator new(unsigned long)>
  b36b32:       0f 1f 40 00             nopl   0x0(%rax)
  b36b36:       66 2e 0f 1f 84 00 00    nopw   %cs:0x0(%rax,%rax,1)
  b36b3d:       00 00 00 
```

Reply:
Yes, we should accept MessageXRef when try to send messages, and return MessageXRef when to dispatch them.